README.md

```diff
@@ -278,3 +278,42 @@ jsonPayload.verb="stacktraceanalyzer"
+
+* The ahead of time compilation can be saved and then loaded for fast startup and restart times on the target hardware.
+
+The tool `train_compile.py` is tightly linked to `train.py` and uses the same configuration file `configs/base.yml`. Here is an example run:
+```
+# Run the below on a single machine, e.g. a v4-8
+python3 MaxText/train_compile.py MaxText/configs/base.yml compile_topology=v5e-256 compile_topology_num_slices=2 \ 
+global_parameter_scale=16 per_device_batch_size=4
+```
+
+This will compile a 16B parameter MaxText model on 2 v5e pods. 
+
+Here is an example that saves then loads the compiled `train_step`, starting with the save:
+```
+# Run the below on a single machine, e.g. a v4-8
+export LIBTPU_INIT_ARGS="--xla_enable_async_all_gather=true" 
+python3 MaxText/train_compile.py MaxText/configs/base.yml compile_topology=v5e-256 \
+compile_topology_num_slices=2 \
+compiled_trainstep_file=my_compiled_train.pickle global_parameter_scale=16 \
+per_device_batch_size=4 steps=10000 learning_rate=1e-3
+```
+
+To load the compiled train_step, you just need to pass `compiled_trainstep_file=my_compiled_train.pickle` into `train.py`:
+```
+# Run the below on each host of the target hardware, e.g. each host on 2 slices of v5e-256
+python3 MaxText/train.py MaxText/configs/base.yml run_name=example_load_compile \
+compiled_trainstep_file=my_compiled_train.pickle \
+global_parameter_scale=16  per_device_batch_size=4 steps=10000 \
+base_output_directory=gs://my-output-bucket dataset_path=gs://my-dataset-bucket
+```
+
+Above we included exporting the compiler flag `LIBTPU_INIT_ARGS` and `learning_rate` because those affect the compiled object `my_compiled_train.pickle.` The sizes of the model (e.g. `global_parameter_scale`, `max_sequence_length` and `per_device_batch`) are fixed when you initally compile via `compile_train.py`, you will see a size error if you try to run the saved compiled object with different sizes than you compiled with. However a subtle note is that the **learning rate schedule** is also fixed when you run `compile_train` - which is determined by both `steps` and `learning_rate`. The optimizier parameters such as  `adam_b1` are passed only as shaped objects to the compiler - thus their real values are determined when you run `train.py`, not during the compilation. If you do pass in different shapes (e.g. `per_device_batch`), you will get a clear error message reporting that the compiled signature has different expected shapes than what was input. If you attempt to run on different hardware than the compilation targets requested via `compile_topology`, you will get an error saying there is a failure to map the devices from the compiled to your real devices. Using different XLA flags or a LIBTPU than what was compiled will silently run with the environment you compiled in without error. 
```

Comment:
That last one should make it clear that there isn't an expected behavior, right? You might get the wrong answer...

Reply:
Good point, clarified.